### ISSUE TITLE
Crash with LogConfusionMatrix Callback with ddp

### ISSUE THREAD

**Issue by lvoegtlin** [ashleve/lightning-hydra-template#189]
Hei everyone,

Thanks a lot for the nice template!
I use it for my experimental setup and it is very nice to use but I have a problem with the LogConfusionMatrix callback when I run on ddp.

If the training reaches a certain length I get at the end of the epoch a 
`UserWarning: resource_tracker: There appear to be 6 leaked semaphore objects to clean up at shutdown`.
Do you have any clue how to solve this problem? I think it's because of the not synced `self.preds` and `self.targets` but I don't know how to synchronize them in python. I also tried to use a metric inside of the callback which backfired hard :D
My log runs with the parameter `rank_zero_only=True`

Thanks a lot for your ideas!

**Comment by lvoegtlin** [comment]
Hei @ashleve 
Thanks for your reply. I will check your link and try to fix it. If I find a solution I will make a PR :)

**Comment by hainanxzg** [comment]
Hi, I am facing the same problem and I saw you open [the issue](https://github.com/PyTorchLightning/pytorch-lightning/discussions/9910) on pytorch-lightning repository. Any improvement now? Can you share your ad-hoc solution?

**Comment by lvoegtlin** [comment]
@hainanxzg I just moved the code from the callback into the LightningModule and trigger it with a boolean parameter. It is ugly but we never had the leaking semaphore problem again.

**Comment by hainanxzg** [comment]
@lvoegtlin Thanks for your reply. I finally find the problem is in my LightningModule's validation_step() or validation_epoch_end().  I think it is something wrong when I try to synchronize the result of all processes of the validation set. Now I use torchmetric to automatically synchronize metrics and my code is：

```
  def validation_step(self, batch: Any, batch_idx: int):
      """
      When training with ddp, this function called on every process.
      If on_epoch is True, the logger automatically logs the end of epoch metric value by calling .compute().
      """
      loss, preds, targets = self.step(batch)

      # log val metrics
      self.val_acc(preds, targets)
      self.val_f1(preds, targets)

      self.log("val/loss", loss, on_step=False, on_epoch=True, prog_bar=True)
      self.log("val/acc", self.val_acc, on_step=False, on_epoch=True, prog_bar=True)
      self.log("val/f1", self.val_f1, on_step=False, on_epoch=True, prog_bar=True)
      return {"loss": loss, "preds": preds, "targets": targets}

  def validation_epoch_end(self, outputs: List[Any]):
      """
      When training with ddp, every process, this function called on every process.
      We want to obtain the metrics of whole validation set, not only on any specific precess.
      However, by default, self.log only
      Note:
      """
      # get best val metrics from current epoch
      acc = self.val_acc.compute()
      self.val_acc_best.update(acc)
      f1 = self.val_f1.compute()
      self.val_f1_best.update(f1)
      self.log("val/acc_best", self.val_acc_best.compute(), on_epoch=True, prog_bar=True)
      self.log("val/f1_best", self.val_f1_best.compute(), on_epoch=True, prog_bar=True)
```
